src/XtensaOptimize.cpp

```diff
@@ -0,0 +1,2331 @@
+    return t.is_uint() && (t.bits() == 32) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<float16_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<float16_t>();
+    return t.is_float() && (t.bits() == 16) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<float>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<float>();
+    return t.is_float() && (t.bits() == 32) && (t.lanes() == vector_size);
+}
+
+bool is_native_vector_type(const Type &t, const Target &target) {
+    int native_lanes = target.natural_vector_size<uint8_t>();
+
+    if (t.is_int_or_uint() && (t.lanes() == native_lanes) && (t.bits() == 8)) {
+        return true;
+    }
+
+    if (t.is_int_or_uint() && (t.lanes() == native_lanes) && (t.bits() == 24)) {
+        return true;
+    }
+
+    if (t.is_int_or_uint() && (t.lanes() == native_lanes / 2) && (t.bits() == 16)) {
+        return true;
+    }
+
+    if (t.is_int_or_uint() && (t.lanes() == native_lanes / 2) && (t.bits() == 48)) {
+        return true;
+    }
+
+    if (t.is_int_or_uint() && (t.lanes() == native_lanes / 4) && (t.bits() == 32)) {
+        return true;
+    }
+
+    if (t.is_float() && (t.lanes() == native_lanes / 4) && (t.bits() == 32)) {
+        return true;
+    }
+
+    return false;
+}
+
+bool is_double_native_vector_type(const Type &t, const Target &target) {
+    int single_vector_bitwidth = 8 * target.natural_vector_size<uint8_t>();
+
+    int double_vector_bitwidth = single_vector_bitwidth * 2;
+    return (t.bits() % 8 == 0) && (double_vector_bitwidth % t.bits() == 0) && (double_vector_bitwidth / t.bits() == t.lanes());
+}
+
+Type get_native_xtensa_vector(const Type &t, const Target &target) {
+    int vector_bitwidth = target.has_feature(Target::Feature::XtensaQ8) ? 1024 : 512;
+    int wide_vector_bitwidth = target.has_feature(Target::Feature::XtensaQ8) ? 4096 : 1536;
+
+    if (t.bits() == 64) {
+        return t.with_lanes(vector_bitwidth / 32);
+    }
+
+    if (t.bits() == 24 || t.bits() == 48) {
+        return t.with_lanes(wide_vector_bitwidth / t.bits());
+    }
+    return t.with_lanes(vector_bitwidth / t.bits());
+}
+
+std::string suffix_for_type(Type t) {
+    if (t.is_bool()) {
+        return "_u1";
+    } else if (t.is_int() && (t.bits() == 8)) {
+        return "_i8";
+    } else if (t.is_uint() && (t.bits() == 8)) {
+        return "_u8";
+    } else if (t.is_int() && (t.bits() == 16)) {
+        return "_i16";
+    } else if (t.is_uint() && (t.bits() == 16)) {
+        return "_u16";
+    } else if (t.is_int() && (t.bits() == 32)) {
+        return "_i32";
+    } else if (t.is_uint() && (t.bits() == 32)) {
+        return "_u32";
+    } else if (t.is_float() && (t.bits() == 32)) {
+        return "_f32";
+    } else if (t.is_float() && (t.bits() == 16)) {
+        return "_f16";
+    }
+
+    return "";
+}
+
+struct Pattern {
+    enum Flags {
```

Comment:
It's not clear what order these flags are supposed to be applied in (i.e. does `NarrowOp0` happen before `SwapOps01` or after? I assume that earlier flags should be processed earliest, but that's not explicitly stated.

Reply:
After looking at the process flags implementation, it seems my intuition was wrong. Thi should probably be documented here.